napari/layers/utils/property_map.py

```diff
@@ -0,0 +1,210 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Generic, Iterable, List, TypeVar
+
+from pydantic import validator
+
+from ...utils import Colormap
+from ...utils.colormaps import ValidColormapArg, ensure_colormap
+from ...utils.events import EventedModel
+from ...utils.events.custom_types import Array
+from .color_transformations import ColorType
+
+OutputType = TypeVar('OutputType')
+
+
+class PropertyMap(ABC, Generic[OutputType], EventedModel):
+    values: List[OutputType] = []
+
+    @abstractmethod
+    def __call__(self, property_row: Dict[str, Any]) -> OutputType:
+        pass
+
+    def refresh(self, properties: Dict[str, Array]):
+        """Updates all values from the given properties.
+
+        Parameters
+        ----------
+        properties : Dict[str, Array]
+            The properties of a layer.
+        """
+        num_values = _num_rows(properties)
+        self.values = self._apply(properties, range(0, num_values))
+
+    def add(self, properties: Dict[str, Array], num_to_add: int):
+        """Adds a number of a new values based on the given properties.
+
+        Parameters
+        ----------
+        properties : Dict[str, Array]
+            The properties of a layer.
+        num_to_add : int
+            The number of values to add.
+        """
+        num_values = _num_rows(properties)
+        indices = range(num_values - num_to_add, num_values)
+        self.values.extend(self._apply(properties, indices))
+
+    def remove(self, indices: Iterable[int]):
+        """Removes some values by index.
+
+        Parameters
+        ----------
+        indices : Iterable[int]
+            The indices to remove.
+        """
+        indices = set(indices)
+        self.values = [
+            self.values[i] for i in range(len(self.values)) if i not in indices
+        ]
+
+    def _apply(self, properties: Dict[str, Array], indices: Iterable[int]):
+        return [
+            self({name: column[index] for name, column in properties.items()})
+            for index in indices
+        ]
+
+    @staticmethod
+    def _num_values(properties):
```

Comment:
`__len__`?

Reply:
Sorry, I should have added an implementation comment here. The intent of this is to return the length of a column in the `properties` table, which should be the same for each column, and should correspond to the number of elements in a layer (e.g. the number of points). That's not generally possible with our dict-of-columns structure because if there are no properties, then we can't know how many elements there are (without more information).

It's used when refreshing `values` in some way (i.e. when it's about to change in size), so `__len__` isn't quite right because I'd expect that to return the length of values.